### PR TITLE
fix(agents): restore compaction gateway logs

### PR DESCRIPTION
**Pull request by rubencu**
## Summary

Describe the problem and fix in 2-5 bullets:

- Problem: live Gateway compaction could happen without any normal Gateway log line showing that compaction started or completed.
- Why it matters: when the embedded agent pauses for compaction, operators watching `pnpm gateway:watch` or `openclaw logs --follow` need to distinguish healthy compaction from a stuck run.
- What changed: compaction start/end now emit info-level embedded-run lifecycle logs with the upstream compaction reason, completion state, retry state, and observed compaction count.
- What did NOT change (scope boundary): no compaction policy, retry behavior, token accounting, config, provider, or hook contract changes.

## Change Type (select all)

- [x] Bug fix
- [ ] Feature
- [ ] Refactor required for the fix
- [x] Docs
- [ ] Security hardening
- [ ] Chore/infra

## Scope (select all touched areas)

- [x] Gateway / orchestration
- [ ] Skills / tool execution
- [ ] Auth / tokens
- [x] Memory / storage
- [ ] Integrations
- [ ] API / contracts
- [x] UI / DX
- [ ] CI/CD / infra

## Linked Issue/PR

- Closes N/A
- Related N/A
- [x] This PR fixes a bug or regression

## Real behavior proof (required for external PRs)

External contributors must show after-fix evidence from a real OpenClaw setup. Unit tests, mocks, lint, typechecks, snapshots, and CI are supplemental only. Screenshots are encouraged even for CLI, console, text, or log changes; terminal screenshots and copied live output count. Be mindful of private information like IP addresses, API keys, phone numbers, non-public endpoints, or other private details when providing evidence.

- Behavior or issue addressed: compaction completed in a live Gateway run, but normal Gateway logs did not expose compaction start/complete lifecycle lines before this patch.
- Real environment tested: local macOS repo checkout, real OpenClaw Gateway process, QA mock provider over HTTP, OpenClaw JSON-RPC `agent` request.
- Exact steps or command run after this patch: `pnpm exec tsx /tmp/openclaw-pr71961-live-compaction.ts <pr-worktree> pr-branch`
- Evidence after fix (screenshot, recording, terminal capture, console output, redacted runtime log, linked artifact, or copied live output):

```json
{
  "label": "pr-branch",
  "compactionEntries": [
    {
      "agentId": "qa",
      "sessionKey": "agent:qa:main",
      "compactionCount": 1
    }
  ],
  "logContainsStart": true,
  "logContainsComplete": true,
  "compactionLogLines": [
    "2026-05-10T11:31:16.082-04:00 [agent/embedded] embedded run auto-compaction start: runId=pr71961-pr-branch-1778427074071-0 reason=threshold",
    "2026-05-10T11:31:16.094-04:00 [agent/embedded] embedded run auto-compaction complete: runId=pr71961-pr-branch-1778427074071-0 reason=threshold compactionCount=1 willRetry=false"
  ],
  "runCount": 1
}
```

- Observed result after fix: the same live Gateway compaction now emits info-level start and complete log lines.
- What was not tested: full repo `pnpm check` / full `pnpm test`; this PR was validated with targeted touched-surface tests, typechecks, live Gateway proof, and Codex review.
- Before evidence (optional but encouraged): same live script against an `origin/main` worktree at the rebase base (`4167d90bc9`) produced `compactionCount: 1` with `logContainsStart: false`, `logContainsComplete: false`, and no compaction log lines.

## Root Cause (if applicable)

For bug fixes or regressions, explain why this happened, not just what changed. Otherwise write `N/A`. If the cause is unclear, write `Unknown`.

- Root cause: the embedded compaction handlers updated session state and fired hooks, but did not write normal info-level lifecycle logs for Gateway operators.
- Missing detection / guardrail: tests covered compaction count and hook behavior, but not the operator-visible log contract.
- Contributing context (if known): existing debug output was not enough for normal Gateway watch/log-follow workflows.

## Regression Test Plan (if applicable)

For bug fixes or regressions, name the smallest reliable test coverage that should catch this. Otherwise write `N/A`.

- Coverage level that should have caught this:
  - [x] Unit test
  - [x] Seam / integration test
  - [ ] End-to-end test
  - [ ] Existing coverage already sufficient
- Target test or file: `src/agents/pi-embedded-subscribe.handlers.compaction.test.ts`, `src/plugins/wired-hooks-compaction.test.ts`
- Scenario the test should lock in: threshold compaction logs start and complete; manual compaction logs start and incomplete; hook wiring passes the upstream compaction reason into the handler.
- Why this is the smallest reliable guardrail: the behavior lives in the embedded compaction handler and hook seam, so these tests cover the new branches without needing provider or UI mocks.
- Existing test that already covers this (if any): existing compaction count/hook tests covered state behavior, not log visibility.
- If no new test is added, why not: N/A

## User-visible / Behavior Changes

Operators now see `embedded run auto-compaction start` and `embedded run auto-compaction complete` in normal Gateway logs when threshold/overflow compaction runs; manual compaction logs as manual compaction.

## Diagram (if applicable)

```text
Before:
[Gateway run crosses compaction threshold] -> [compaction happens] -> [no normal lifecycle log]

After:
[Gateway run crosses compaction threshold] -> [start log] -> [compaction happens] -> [complete/incomplete log]
```

## Security Impact (required)

- New permissions/capabilities? (`Yes/No`) No
- Secrets/tokens handling changed? (`Yes/No`) No
- New/changed network calls? (`Yes/No`) No
- Command/tool execution surface changed? (`Yes/No`) No
- Data access scope changed? (`Yes/No`) No
- If any `Yes`, explain risk + mitigation: N/A

### Environment

- OS: macOS
- Runtime/container: Node 25.2.1 for local commands; repo runtime remains Node 22+ compatible
- Model/provider: QA mock provider
- Integration/channel (if any): Gateway JSON-RPC `agent`
- Relevant config (redacted): QA agent with small test compaction budget (`contextWindow: 4000`, `reserveTokens: 3950`, `keepRecentTokens: 1`), Gateway logging at `info`, memory flush disabled for the live proof

### Steps

1. Start a real Gateway process from the repo under test.
2. Start the QA mock provider and send a JSON-RPC `agent` request with a prompt large enough to trigger threshold compaction.
3. Inspect the live Gateway log output and session store compaction count.

### Expected

- The session records `compactionCount: 1`.
- Gateway logs include compaction start and complete lifecycle lines.

### Actual

- Before this patch on `4167d90bc9`: `compactionCount: 1`, but no compaction start or complete log lines.
- After this patch on `6ee42aaa4e`: `compactionCount: 1`, plus start and complete log lines.

## Evidence

Attach at least one:

- [x] Failing test/log before + passing after
- [x] Trace/log snippets
- [ ] Screenshot/recording
- [ ] Perf numbers (if relevant)

Validation run:

```text
pnpm exec oxfmt --check --threads=1 docs/concepts/compaction.md docs/reference/session-management-compaction.md src/agents/pi-embedded-subscribe.handlers.compaction.ts src/agents/pi-embedded-subscribe.handlers.compaction.test.ts src/agents/pi-embedded-subscribe.handlers.ts src/agents/pi-embedded-subscribe.handlers.types.ts src/agents/openclaw-owned-tool-runtime-contract.test.ts src/agents/pi-embedded-subscribe.handlers.messages.test.ts src/agents/pi-embedded-subscribe.handlers.tools.media.test.ts src/agents/pi-embedded-subscribe.handlers.tools.test.ts src/agents/pi-tool-definition-adapter.after-tool-call.fires-once.test.ts src/plugins/wired-hooks-after-tool-call.e2e.test.ts src/plugins/wired-hooks-compaction.test.ts
git diff --check
pnpm test src/agents/pi-embedded-subscribe.handlers.compaction.test.ts
pnpm test src/agents/pi-embedded-subscribe.handlers.lifecycle.test.ts
pnpm test src/agents/pi-embedded-subscribe.handlers.compaction.test.ts src/agents/pi-embedded-subscribe.subscribe-embedded-pi-session.waits-multiple-compaction-retries-before-resolving.test.ts src/agents/pi-embedded-subscribe.subscribe-embedded-pi-session.splits-long-single-line-fenced-blocks-reopen.test.ts -- -t compaction
pnpm test src/agents/pi-embedded-subscribe.handlers.compaction.test.ts src/agents/pi-embedded-subscribe.handlers.lifecycle.test.ts src/plugins/wired-hooks-compaction.test.ts src/agents/openclaw-owned-tool-runtime-contract.test.ts src/agents/pi-embedded-subscribe.handlers.tools.test.ts src/plugins/wired-hooks-after-tool-call.e2e.test.ts src/agents/pi-tool-definition-adapter.after-tool-call.fires-once.test.ts src/agents/pi-embedded-subscribe.handlers.messages.test.ts src/agents/pi-embedded-subscribe.handlers.tools.media.test.ts
pnpm tsgo:core
pnpm tsgo:test:src
codex review --base origin/main
```

Codex review result: no actionable correctness issues found.

## Human Verification (required)

What you personally verified (not just CI), and how:

- Verified scenarios: live Gateway before/after proof, threshold compaction start/complete logs, manual compaction incomplete branch, hook wiring, compaction count persistence, touched type surfaces.
- Edge cases checked: missing compaction result logs as incomplete; retry/debug path remains covered by existing compaction tests; no `CHANGELOG.md` diff remains in this PR.
- What you did **not** verify: full repo test/check sweep, maintainer-owned manual verification. User manual verification: owner performs final manual verification outside this agent run.

## Review Conversations

- [x] I replied to or resolved every bot review conversation I addressed in this PR.
- [x] I left unresolved only the conversations that still need reviewer or maintainer judgment.

## Compatibility / Migration

- Backward compatible? (`Yes/No`) Yes
- Config/env changes? (`Yes/No`) No
- Migration needed? (`Yes/No`) No
- If yes, exact upgrade steps: N/A

## Risks and Mitigations

- Risk: log output becomes noisier for sessions that compact repeatedly.
  - Mitigation: the patch emits one start line and one end line per compaction lifecycle at info level, with no extra per-token or streaming logs.

### CODE REVIEW

File: src/agents/pi-embedded-subscribe.handlers.compaction.test.ts [hunk 1]
```diff
@@ -113,6 +116,60 @@ describe("reconcileSessionStoreCompactionCountAfterSuccess", () => {
...
+        event: "embedded_run_compaction_end",
+        reason: "threshold",
+        runId: "run-test",
+        completed: true,
+        compactionCount: 1,
+        consoleMessage:
+          "embedded run auto-compaction complete: runId=run-test reason=threshold compactionCount=1 willRetry=false",
+      }),
+    );
+  });
```
[the remark this answers is not in the document]
Comment (rubencu) [reply src/agents/pi-embedded-subscribe.handlers.compaction.test.ts]: Addressed by hoisting the lifecycle logging test into its own `describe("compaction lifecycle logging", ...)` block. Verified with `pnpm test src/agents/pi-embedded-subscribe.handlers.compaction.test.ts` and `OPENCLAW_VITEST_MAX_WORKERS=1 pnpm check:changed`.

**Comment by rubencu** [comment]
@codex review

**Comment by steipete** [comment]
Landed via rebase onto `main`.

- Source head: https://github.com/openclaw/openclaw/commit/74dab5b33181eba6b5048535fea7fa37053f5f7c
- Landed commit: https://github.com/openclaw/openclaw/commit/fd38a8a36dabe730837dda28600eea2085810ccf
- Merge commit reported by GitHub: https://github.com/openclaw/openclaw/commit/fd38a8a36dabe730837dda28600eea2085810ccf
- Gate: GitHub CI rollup green on `74dab5b33181eba6b5048535fea7fa37053f5f7c`
- Extra proof: `pnpm test src/agents/pi-embedded-subscribe.handlers.compaction.test.ts`; `pnpm check:changed` on Blacksmith Testbox `tbx_01kranysc0aw98n9k48n33nm7s`; `pnpm check:test-types`; `pnpm test src/plugins/bundle-manifest.test.ts`; `pnpm test src/plugins/provider-validation.test.ts`
- Duplicate search: no separate duplicate found beyond this PR.

Thanks @rubencu!
